### PR TITLE
[DT-041410] Project level OAuth tokens are unavailable for use in module steps

### PR DESCRIPTION
**Pull request by rayansami**

### CODE REVIEW

File: Decisions.Microsoft365.Exchange/Steps/CalendarSteps.cs [hunk 1]
```diff
@@ -17,54 +17,54 @@ public class CalendarSteps
         };
         
         public Microsoft365Event? CreateCalendarEvent(Microsoft365CalendarEvent microsoft365CalendarEvent, string userIdentifier, string? calendarId,
-            [PropertyClassification(0, "Settings Override", "Settings")] ExchangeSettings? settingsOverride)
+            [PropertyClassification(0, "Settings Override", "Settings")] InputExchangeSettings? settingsOverride)
```
Comment (noah-decisions) [inline Decisions.Microsoft365.Exchange/Steps/CalendarSteps.cs]: I would test what happens for an upgrade here. We may actually have to make new methods for all of these steps to be able to update the input without breaking people already using the steps

File: Decisions.Microsoft365.Exchange/ExchangeSettings.cs [hunk 2]
```diff
@@ -12,6 +13,8 @@
 using DecisionsFramework.ServiceLayer.Services.Administration;
 using DecisionsFramework.ServiceLayer.Services.Folder;
 using DecisionsFramework.ServiceLayer.Utilities;
+using DecisionsFramework.Utilities.Data.ObjectComparison;
+using Newtonsoft.Json;
```
Comment (jason-decisions) [inline Decisions.Microsoft365.Exchange/ExchangeSettings.cs]: Why are we adding imports but not changing anything. I think this file has been included by mistake.

File: Decisions.Microsoft365.Exchange/InputExchangeSettings.cs [hunk 3]
```diff
@@ -0,0 +1,55 @@
+using System.ComponentModel;
+using DecisionsFramework.Design.ConfigurationStorage.Attributes;
+using DecisionsFramework.Design.Properties;
+using DecisionsFramework.Design.Properties.Attributes;
+
+namespace Decisions.Microsoft365.Exchange;
+
+public class InputExchangeSettings: ICreateChildEntityInFolder
```
Comment (jason-decisions) [inline Decisions.Microsoft365.Exchange/InputExchangeSettings.cs:8]: I'm really confused. I think you've branched off the wrong place or something. What are these changes?

File: Decisions.Microsoft365.Exchange/Steps/EmailSteps.cs [hunk 4]
```diff
@@ -19,99 +18,53 @@ public class EmailSteps
             return JsonHelper<Microsoft365Message?>.JsonDeserialize(result);
         }
         
-        public Microsoft365Message?[] SearchEmails(string userIdentifier, string searchQuery, int? maxPageCount,
+        public Microsoft365EmailList? SearchEmails(string userIdentifier, string searchQuery,
```
Comment (jason-decisions) [inline Decisions.Microsoft365.Exchange/Steps/EmailSteps.cs]: Why are we modifying this?

File: Decisions.Microsoft365.Exchange/Steps/EmailSteps.cs [hunk 5]
```diff
@@ -19,99 +18,53 @@ public class EmailSteps
...
-            List<Microsoft365Message>? messages = new List<Microsoft365Message>();
-            foreach (Microsoft365EmailList? emailList in emailLists)
-            {
-                foreach (Microsoft365Message email in emailList?.Value!)
-                {
-                    messages.Add(email);
-                }
-            }
-
-            return messages.ToArray();
```
Comment (jason-decisions) [inline Decisions.Microsoft365.Exchange/Steps/EmailSteps.cs]: How are these changes relevant here?

File: Decisions.Microsoft365.Exchange/Steps/EmailSteps.cs (hunk 4 above)
Comment (rayansami) [reply Decisions.Microsoft365.Exchange/Steps/EmailSteps.cs]: There shouldn't be any changes here, other than removed _AutoRegister_ Attr. Fixed

File: Decisions.Microsoft365.Exchange/Steps/EmailSteps.cs (hunk 5 above)
Comment (rayansami) [reply Decisions.Microsoft365.Exchange/Steps/EmailSteps.cs]: Fixed.

File: Decisions.Microsoft365.Exchange/ExchangeSettings.cs (hunk 2 above)
Comment (rayansami) [reply Decisions.Microsoft365.Exchange/ExchangeSettings.cs]: Removed the unnecessary imports. These got added when I changed some code here. Fixed it

File: Decisions.Microsoft365.Exchange/InputExchangeSettings.cs (hunk 3 above)
Comment (rayansami) [reply Decisions.Microsoft365.Exchange/InputExchangeSettings.cs:8]: This is the new input to the duplicate steps (CalendarSteps -> ExchangeCalendarSteps and so on)
That what we were discussing earlier. SInce ExchangeSettings inherits AbstractModuleSettings, there can be only one instance of it per project. So instead of using ExchangeSettings, we introduced InputExchangeSettings and pass it to the methods(steps) instead of passing ExchangeSettings. 

tagging @noah-decisions if I missed more context.

Comment (noah-decisions) [reply Decisions.Microsoft365.Exchange/InputExchangeSettings.cs:8]: Yes, this is the new object to use for the settings override input to the steps, replacing where we used the modules settings object before

File: Decisions.Microsoft365.Exchange/Decisions.Microsoft365.Exchange.csproj [hunk 6]
```diff
@@ -14,7 +14,7 @@
   </ItemGroup>
 
   <ItemGroup>
-    <PackageReference Include="DecisionsSDK" Version="8.19.0" />
+    <PackageReference Include="DecisionsSDK" Version="9.0.0" />
```
Comment (jason-decisions) [inline Decisions.Microsoft365.Exchange/Decisions.Microsoft365.Exchange.csproj]: Please update to latest SDK (9.1.0 - published Friday)

File: Decisions.Microsoft365.Exchange/ExchangeSettings.cs
Comment (jason-decisions) [inline Decisions.Microsoft365.Exchange/ExchangeSettings.cs:1]: I think this class goes away right? And the new class becomes the Exchange Settings object?

File: Decisions.Microsoft365.Exchange/InputExchangeSettings.cs (hunk 3 above)
Comment (jason-decisions) [inline Decisions.Microsoft365.Exchange/InputExchangeSettings.cs:8]: If this becomes the new Exchange Settings Object, I think we need the ORM Entity attribute that we had in `ExchangeSettings.cs`, right?

File: Decisions.Microsoft365.Exchange/InputExchangeSettings.cs [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+{
+    private string graphUrl = "https://graph.microsoft.com/v1.0";
+
+    [PropertyClassification(0, "Graph URL", "Exchange Settings")]
+    [WritableValue]
+    public string GraphUrl
```
Comment (jason-decisions) [inline Decisions.Microsoft365.Exchange/InputExchangeSettings.cs:14]: Once we update this to be an `ORMEntity`, we need to add `DataMember` attributes to all of these.

File: Decisions.Microsoft365.Exchange/ExchangeSettings.cs
Comment (noah-decisions) [reply Decisions.Microsoft365.Exchange/ExchangeSettings.cs:1]: No the new class is designed to be used as the input on the steps for overriding the settings. This is still the settings object for the module

Comment (rayansami) [reply Decisions.Microsoft365.Exchange/ExchangeSettings.cs:1]: This class stays so that the users who already have the steps implemented, works as usual on upgrade. But going forward, when they implement the steps, they are going to use new object (_InputExchangeSettings_ instead of _ExchangeSettings_)

File: Decisions.Microsoft365.Exchange/InputExchangeSettings.cs (hunk 3 above)
Comment (noah-decisions) [reply Decisions.Microsoft365.Exchange/InputExchangeSettings.cs:8]: The intent behind this is more like the SmtpServerSettings object. The step will hold an instance of this object unique to the step it's configured on, so it wouldn't be an ORM entity here

Comment (rayansami) [reply Decisions.Microsoft365.Exchange/InputExchangeSettings.cs:8]: Discussed this with @noah-decisions, basically we are overriding the inputs using InputExchangeSettings, just like it's being done on SmtpServerSettings (not an ORM entity) that overrides SMTP Settings (which are actually this object as a property of the settings object).

ExchangeSettings is still the settings object for the module and InputExchangeSettings is just overriding the input to the steps. If the user doesn't define it, the steps should use the settings saved to the other(ExchangeSettings) object. Making our new class(InputExchangeSettings ) an ORM object would just cause more issues

File: Decisions.Microsoft365.Exchange/Decisions.Microsoft365.Exchange.csproj (hunk 6 above)
Comment (rayansami) [reply Decisions.Microsoft365.Exchange/Decisions.Microsoft365.Exchange.csproj]: Updated

File: Decisions.Microsoft365.Exchange/ExchangeSettings.cs
Comment (jason-decisions) [reply Decisions.Microsoft365.Exchange/ExchangeSettings.cs:1]: If we're entirely using step level settings that duplicate all of this code, then where are these settings ever used?

Comment (noah-decisions) [reply Decisions.Microsoft365.Exchange/ExchangeSettings.cs:1]: The GetSettings method of the GraphRest class gets the regular module settings, then maps values from the override if they are not null or empty.

@rayansami you should add logic in your steps to either pass a null ExchangeSettings when relevant (settingsOverride is null or both properties are null/empty), or add '?' to where you are accessing the properties from settingsOverride so that a null input won't throw an exception, and the resulting ExchangeSettings object will have null on both fields, meaning the stored module settings will be used.

Comment (rayansami) [reply Decisions.Microsoft365.Exchange/ExchangeSettings.cs:1]: Updated with null check on accessing InputExchangeSettings properties on steps

**Review by jason-decisions — APPROVED** [review]
